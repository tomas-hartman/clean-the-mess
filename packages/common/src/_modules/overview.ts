import { Tabs } from 'webextension-polyfill';
import { Overview, OverviewItem } from '../popup';
import { getHash } from './helpers';

const getOriginUrl = (tabData: Tabs.Tab) => {
  if (!tabData.url) return "Other tabs";

  const url = new URL(tabData.url);

  if (!(url.protocol === "http:" || url.protocol === "https:")) {
    switch (url.protocol) {
      case 'about:':
      case 'moz-extension:':
      case 'chrome:':
        return 'Browser tabs';
      case 'file:':
        return 'Opened files';
      default:
        return 'Other tabs';
    }
  }

  if (url.hostname === 'localhost') {
    return 'Localhost';
  }

  if (url.origin) {
    return url.origin
  }

  return 'Other tabs';
};

/**
 * Function that creates data structure for overview grouping.
 * @todo It also handles group naming, but it needs refactoring.
 * @param {Object[]} tabs - Standard tabs object from browser
 * @returns {Object[]} Sorted array of objects that are used for overview grouping
 */
export const getOverview = (tabs: Tabs.Tab[]): Overview => {
  const output: Required<OverviewItem>[] = [];

  tabs.forEach((tab) => {
    const originUrl = getOriginUrl(tab);

    if (!tab.pinned && tab.id !== undefined) {
      if (output.some((website) => website.url === originUrl)) {
        const index = output.findIndex(
          (website) => website.url === originUrl,
        );

        output[index].count += 1;
        output[index].ids.push(tab.id);
      } else {
        const key = getHash(originUrl);

        output.push({
          url: originUrl,
          count: 1,
          ids: [tab.id],
          key,
          favicon: tab.favIconUrl,
        });
      }
    }
  });

  output.sort((a, b) => b.count - a.count);

  return output;
};
