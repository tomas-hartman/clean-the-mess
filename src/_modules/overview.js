import { getHash } from './helpers';

const getOriginUrl = (tabData) => {
  const url = new URL(tabData.url);
  let originUrl = '';

  try {
    originUrl = url.origin;

    if (originUrl === 'null' || url.protocol === 'moz-extension:' || url.protocol === 'chrome:' || url.protocol === 'file:') {
      switch (url.protocol) {
        case 'about:':
        case 'moz-extension:':
        case 'chrome:':
          originUrl = 'Browser tabs';
          break;
        case 'file:':
          originUrl = 'Opened files';
          break;
        case 'localhost:':
          originUrl = 'Localhost';
          break;
        default:
          originUrl = 'Other tabs';
          break;
      }
    }
  } catch (err) {
    if (tabData.url === 'localhost') {
      originUrl = 'Localhost';
    } else if ((/^((\d{1,3}.){3}\d{1,3})(:|\/|\s|$)/g).test(tabData.url)) {
      const array = tabData.url.split(/:|\//);
      [originUrl] = array;
    } else {
      originUrl = 'Other tabs';
    }
  }

  return originUrl;
};

/**
 * Function that creates data structure for overview grouping.
 * @todo It also handles group naming, but it needs refactoring.
 * @param {Object[]} tabs - Standard tabs object from browser
 * @returns {Object[]} Sorted array of objects that are used for overview grouping
 */
export const getOverview = (tabs) => {
  const output = [];

  tabs.forEach((tab) => {
    const originUrl = getOriginUrl(tab);

    if (!tab.pinned) {
      if (output.some((website) => website.url === originUrl)) {
        const index = output.findIndex(
          (website) => website.url === originUrl,
        );

        output[index].count += 1;
        output[index].ids.push(tab.id);
      } else {
        const key = getHash(originUrl);

        output.push({
          url: originUrl,
          count: 1,
          ids: [tab.id],
          key,
          favicon: tab.favIconUrl,
        });
      }
    }
  });

  output.sort((a, b) => b.count - a.count);

  // browser.browserAction.setBadgeText({text: `${tabs.length}`});

  return output;
};
