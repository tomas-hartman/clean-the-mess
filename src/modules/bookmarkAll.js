import { isSupportedProtocol, getTabDataFromId } from './helpers.js';

const addBookmarksToFolder = async (tabIds, parentId) => {
	for(let i = 0; i < tabIds.length; i++) {
		const [title, url] = await getTabDataFromId(tabIds[i]);
    
		/**
		 * This does not search only in the given folder but everywhere!
		 * Currently a feature (to keep bookmarks clean), but it could be reconsidered.
		 */
		const duplicates = await browser.bookmarks.search(url); 

		if(duplicates.length === 0 && isSupportedProtocol(url)){
			await browser.bookmarks
				.create({ title, url, parentId })
				.catch((err) => console.log(err));
		} else console.log(`Item with url ${url} already exists and was skipped.`);
	}
};

/**
 * Handles merge in bookmarks folder and duplicate folder/bookmarks names.
 * 
 * @param {array} tabIds Array of ids of current tabs
 * @param {array} foundItems Array of items already found in bookmarks (usually should be 0 or 1)
 * @param {string} folderName Name of new folder in case it should be created
 */
const mergeBookmarksInFolder = async (tabIds, foundItems, folderName) => {
	// Returns first folder of its name into which I will add the new files
	const folder = foundItems.find((item) => item.type === 'folder');

	if(!folder) {
		const newFolder = await browser.bookmarks.create({title: folderName });
		await addBookmarksToFolder(tabIds, newFolder.id);

		console.log('Items added to new folder!');

		return;
	} 

	const parentId = folder.id;
	const folderChildren = await browser.bookmarks.getChildren(parentId);

	for(let i=0;i<tabIds.length;i++){
		const [title, url] = await getTabDataFromId(tabIds[i]);

		if(!isSupportedProtocol(url)) return;

		const isAlreadyBookmarked = folderChildren.some((child) => child.title === title);

		if(!isAlreadyBookmarked) {
			await browser.bookmarks
				.create({ title, url, parentId })
				.catch((err) => console.log(err));

			console.log('Items were merged with those from folder.');
		} else console.log(`Item with url ${url} is already in folder ${folderName} and was skipped.`);
	}
};

/**
 * Handles bookmark-all calls.
 * 
 * 1. assumes folder-name and checks if given folder name already exists
 * 2. if not, it creates the folder and adds all of them in the folder
 * 3. if so, it merges new items with the items already found in the folder
 * 
 * Duplicate bookmarks are always skipped. In case of duplicate folders, first folder is used.
 * 
 * @param {Object} _data Data object containing tabsOverviewId  
 * @param {Object} _data.overviewObject tabsOverview item
 * @param {number} _data.index tabsOverview id
 */
export const handleBookmarkAll = async (_data) => {
	const { overviewObject, index } = _data;
	const {bookmarks} = browser;

	// It does not merge items in bookmark root
	if(overviewObject.ids.length === 1) {
		await addBookmarksToFolder(overviewObject.ids, null);
		browser.runtime.sendMessage({ type: 'items-bookmarked', data: { index }});

		return;
	}

	const folderName = (new URL(overviewObject.url)).hostname;
	const searchResults = await bookmarks.search({title: folderName});

	if (await searchResults.length === 0) {
		const newFolder = await bookmarks.create({title: folderName });
        
		await addBookmarksToFolder(overviewObject.ids, newFolder.id);

		console.log('Items added to new folder!');
	} else {
		await mergeBookmarksInFolder(overviewObject.ids, searchResults, folderName);
	}

	browser.runtime.sendMessage({ type: 'items-bookmarked', data: { index }});
};