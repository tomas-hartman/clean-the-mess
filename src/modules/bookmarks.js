import { isSupportedProtocol, getTabDataFromId, hasIgnoredProtocol } from './helpers.js';

/**
 * Background
 */

const bookmarkTabsToFolder = async (tabIds, parentId) => {
	for(let i = 0; i < tabIds.length; i++) {
		const [title, url] = await getTabDataFromId(tabIds[i]);
    
		/**
		 * This does not search only in the given folder but everywhere!
		 * Currently a feature (to keep bookmarks clean), but it could be reconsidered.
		 */
		const duplicates = await browser.bookmarks.search(url); 

		if(duplicates.length === 0 && isSupportedProtocol(url)){
			await browser.bookmarks
				.create({ title, url, parentId })
				.catch((err) => console.log(err));
		} else console.log(`Item with url ${url} already exists and was skipped.`);
	}
};

/**
 * Handles merge in bookmarks folder and duplicate folder/bookmarks names.
 * 
 * @param {array} tabIds Array of ids of current tabs
 * @param {array} foundItems Array of items already found in bookmarks (usually should be 0 or 1)
 * @param {string} folderName Name of new folder in case it should be created
 */
const mergeBookmarksInFolder = async (tabIds, foundItems, folderName) => {
	// Returns first folder of its name into which I will add the new files
	const folder = foundItems.find((item) => item.type === 'folder');

	if(!folder) {
		const newFolder = await browser.bookmarks.create({title: folderName });
		await bookmarkTabsToFolder(tabIds, newFolder.id);

		console.log('Items added to new folder!');

		return;
	} 

	const parentId = folder.id;
	const folderChildren = await browser.bookmarks.getChildren(parentId);

	for(let i=0;i<tabIds.length;i++){
		const [title, url] = await getTabDataFromId(tabIds[i]);

		if(!isSupportedProtocol(url)) return;

		const isAlreadyBookmarked = folderChildren.some((child) => child.title === title);

		if(!isAlreadyBookmarked) {
			await browser.bookmarks
				.create({ title, url, parentId })
				.catch((err) => console.log(err));

			console.log('Items were merged with those from folder.');
		} else console.log(`Item with url ${url} is already in folder ${folderName} and was skipped.`);
	}
};

/**
 * Handles bookmark-all calls.
 * 
 * 1. assumes folder-name and checks if given folder name already exists
 * 2. if not, it creates the folder and adds all of them in the folder
 * 3. if so, it merges new items with the items already found in the folder
 * 
 * Duplicate bookmarks are always skipped. In case of duplicate folders, first folder is used.
 * 
 * @param {Object} _data Data object containing tabsOverviewId  
 * @param {Object} _data.overviewObject tabsOverview item
 * @param {number} _data.index tabsOverview id
 */
export const handleBookmarkAll = async (_data) => {
	const { overviewObject, index } = _data;
	const {bookmarks} = browser;

	// It does not merge items in bookmark root
	if(overviewObject.ids.length === 1) {
		await bookmarkTabsToFolder(overviewObject.ids, null);
		browser.runtime.sendMessage({ type: 'items-bookmarked', data: { index }});

		return;
	}

	const folderName = (new URL(overviewObject.url)).hostname;
	const searchResults = await bookmarks.search({title: folderName});

	if (await searchResults.length === 0) {
		const newFolder = await bookmarks.create({title: folderName });
        
		await bookmarkTabsToFolder(overviewObject.ids, newFolder.id);

		console.log('Items added to new folder!');
	} else {
		await mergeBookmarksInFolder(overviewObject.ids, searchResults, folderName);
	}

	browser.runtime.sendMessage({ type: 'items-bookmarked', data: { index }});
};


/**
 * Popup
 */

/**
 * Adds bookmark status to given element
 * @param {Object} item Object from screen detailed object
 * @param {Element} _document instance of document in whose context is bookmarked element located
 */
export const addBookmarkStatus = async (item, _document) => {
	// I will not check for duplicate items with some special protocols in bookmarks
	if(hasIgnoredProtocol(item.url)) return;

	const bookmarks = await browser.bookmarks.search({ url: item.url });
	const elm = _document.querySelector(`li[data-tab-id='${item.id}'] .bookmark`);

	if (bookmarks.length > 0) {
		elm.classList.add('bookmarked');
		elm.classList.remove('bookmark-close');
		elm.setAttribute('title', 'This url is already bookmarked');
	}
};

/**
 * Saves element with given _id as a new bookmark with given title and url and plays browser action with icon
 * @param {string} url 
 * @param {string} title 
 * @param {number} _id 
 */
export const bookmarkTab = async (url, title, _id) => {
	if (isSupportedProtocol(url)) {
		// check if folder exists and create special folder?
		await browser.bookmarks
			.create({ title: title, url: url })
			.catch((err) => console.log(err));
		await browser.browserAction.setIcon({
			path: '../icons/btn-bookmark-star-blue.svg',
		});

		new Promise((resolve, _reject) => {
			setTimeout(() => {
				browser.browserAction.setIcon({});
				resolve();
			}, 700);
		});
		console.log('New bookmark created');
	}
};