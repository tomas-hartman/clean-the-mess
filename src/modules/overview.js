import { getHash } from './helpers.js';

/**
 * Function that creates data structure for overview grouping.
 * @todo It also handles group naming, but it needs refactoring.
 * @param {Object[]} tabs - Standard tabs object from browser
 * @returns {Object[]} Sorted array of objects that are used for overview grouping
 */
const getOverview = (tabs) => {
  const output = [];

  tabs.forEach((tab) => {
    let url = {};
    let originUrl = '';

    try {
      url = new URL(tab.url);
      originUrl = url.origin;

      if (originUrl === 'null' || url.protocol === 'moz-extension:' || url.protocol === 'chrome:' || url.protocol === 'file:') {
        switch (url.protocol) {
          case 'about:':
          case 'moz-extension:':
          case 'chrome:':
            originUrl = 'Browser tabs';
            break;
          case 'file:':
            originUrl = 'Opened files';
            break;
          case 'localhost:':
            originUrl = 'Localhost';
            break;
          default:
            originUrl = 'Other tabs';
            break;
        }
      }
    } catch (err) {
      if (tab.url === 'localhost') {
        originUrl = 'Localhost';
      } else if ((/^((\d{1,3}.){3}\d{1,3})(:|\/|\s|$)/g).test(tab.url)) {
        const array = tab.url.split(/:|\//);
        originUrl = array[0];
      } else {
        originUrl = 'Other tabs';
      }
    }

    if (!tab.pinned) {
      if (output.some((website) => website.url === originUrl)) {
        const index = output.findIndex(
          (website) => website.url === originUrl,
        );

        output[index].count += 1;
        output[index].ids.push(tab.id);
      } else {
        const key = getHash(originUrl);
        output.push({
          url: originUrl, count: 1, ids: [tab.id], key,
        });
      }
    }
  });

  output.sort((a, b) => b.count - a.count);

  // browser.browserAction.setBadgeText({text: `${tabs.length}`});

  return output;
};

export default getOverview;
