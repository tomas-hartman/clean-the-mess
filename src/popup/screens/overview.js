/* 
1. header
2. list

returns whole screen
*/