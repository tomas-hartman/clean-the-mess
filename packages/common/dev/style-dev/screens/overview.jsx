import ReactDOM from 'react-dom';

// import '../style.scss';

import { OverviewScreen } from '../../../src/popup/screens/Overview/OverviewScreen';
import { getOverview } from '../../../src/_modules';
import { tabs } from '../samples/firefox/tabs';

const tabsOverview = getOverview(tabs);

ReactDOM.render(
  <OverviewScreen
    overviewData={tabsOverview}
    headerData={{ openTabs: 10 }}
    className="slide-in"
    switchToScreen={() => true}
    closeTabs={() => true}
  />,
  document.getElementsByClassName('dev-wrapper-overview')[0],
);
