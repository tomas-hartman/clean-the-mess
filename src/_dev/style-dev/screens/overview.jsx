import ReactDOM from 'react-dom';

// import '../style.scss';

import { OverviewScreen } from '../../../popup/screens/Overview/OverviewScreen';
import { getOverview } from '../../../_modules';
import { tabs } from '../samples/firefox/tabs';

const tabsOverview = getOverview(tabs);

ReactDOM.render(
  <OverviewScreen
    overviewData={tabsOverview}
    headerData={{ openTabs: 10 }}
    className="slide-in"
    switchToScreen={() => true}
    closeTabs={() => true}
  />,
  document.getElementsByClassName('dev-wrapper-overview')[0],
);
