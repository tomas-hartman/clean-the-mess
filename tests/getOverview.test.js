import { getOverview } from '../src/modules/overview';
import { tabs } from './__sampleData__/tabs';

it('should return correct overview data', () => {
  const output = getOverview(tabs);

  expect(output).toMatchSnapshot();
});

it('overview should have valid properties', () => {
  const sampleMinifiedTabs = [
    {
      id: 5,
      index: 0,
      windowId: 1,
      pinned: false,
      status: 'complete',
      lastAccessed: 1589322068715,
      audible: false,
      mutedInfo: {
        muted: false,
      },
      sharingState: {
        camera: false,
        microphone: false,
      },
      successorTabId: -1,
      url: 'https://www.google.com/?q=1',
      title: 'Google jedna',
    },
    {
      id: 7,
      index: 1,
      windowId: 1,
      pinned: false,
      status: 'complete',
      lastAccessed: 1589831142586,
      audible: false,
      mutedInfo: {
        muted: false,
      },
      sharingState: {
        camera: false,
        microphone: false,
      },
      successorTabId: -1,
      url: 'https://www.google.com/?q=2',
      title: 'Google 2',
    },
    {
      id: 15,
      index: 2,
      windowId: 1,
      pinned: false,
      status: 'complete',
      lastAccessed: 1589831456789,
      audible: false,
      mutedInfo: {
        muted: false,
      },
      sharingState: {
        camera: false,
        microphone: false,
      },
      successorTabId: -1,
      url: 'https://www.example.com/?q=1',
      title: 'Example 1',
    },
  ];
  const output = getOverview(sampleMinifiedTabs);
  const item = output[0];

  expect(output).toBeInstanceOf(Array);
  expect(output.length).toBe(2); // because 2 x google and 1 x example

  expect(item).toBeInstanceOf(Object);
  expect(item).toHaveProperty('url');
  expect(item).toHaveProperty('count');
  expect(item).toHaveProperty('ids');

  expect(item.ids).toBeInstanceOf(Array);
  expect(item.ids.length).toBe(item.count);

  expect(item.url).toBe('https://www.google.com');
});

describe('Output of url/title', () => {
  it.each([
    ['about:', { url: 'about:debugging', title: 'about:debugging' }, 'Browser tabs'],
    ['moz-extension:', { url: 'moz-extension://58ee5ecd-6e4e-4ff4-8d70-56bca7d26444', title: 'manifest.json' }, 'Browser tabs'],
    ['chrome:', { url: 'chrome://extensions/', title: 'chrome://extensions/' }, 'Browser tabs'],
    ['file:', { url: 'file:///C:/clean-the-mess/index.html', title: 'Test report' }, 'Opened files'],
    ['ftp:', { url: 'ftp://www.example.com/index.html', title: 'Test report' }, 'ftp://www.example.com'],
    ['localhost', { url: 'localhost', title: 'localhost' }, 'Localhost'],
    ['localhost:3000', { url: 'localhost:3000', title: 'localhost:3000' }, 'Localhost'],
    ['IP', { url: '127.0.0.1', title: 'IP address' }, '127.0.0.1'],
    ['IP:port', { url: '127.0.0.1:3000', title: 'IP address with port' }, '127.0.0.1'],
    // Failcases:
    ['random string', { url: 'random string', title: 'random string' }, 'Other tabs'],
    ['empty string', { url: '', title: '' }, 'Other tabs'],
  ])(
    'output should work fine with non-urls - %s',
    (_tcTitle, testData, expected) => {
      const { url, title } = testData;
      const sampleMinifiedTabs = [
        {
          id: 5,
          index: 0,
          windowId: 1,
          pinned: false,
          status: 'complete',
          lastAccessed: 1589322068712,
          url,
          title,
        },
      ];

      const output = getOverview(sampleMinifiedTabs);
      const item = output[0];

      expect(item.url).toBe(expected);
    },
  );
});

it('should skip pinned items', () => {
  const sampleMinifiedTabs = [
    {
      id: 6,
      index: 0,
      windowId: 1,
      pinned: true,
      status: 'complete',
      lastAccessed: 1589322068733,
      url: 'https://www.google.com',
      title: 'Google.com but pinned',
    },
  ];

  const output = getOverview(sampleMinifiedTabs);

  expect(output.length).toBe(0);
});
