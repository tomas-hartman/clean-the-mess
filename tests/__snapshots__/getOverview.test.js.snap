// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should return correct overview data 1`] = `
Array [
  Object {
    "count": 5,
    "ids": Array [
      77,
      80,
      92,
      97,
      99,
    ],
    "url": "https://www.google.com",
  },
  Object {
    "count": 5,
    "ids": Array [
      87,
      95,
      94,
      98,
      100,
    ],
    "url": "https://stackoverflow.com",
  },
  Object {
    "count": 3,
    "ids": Array [
      81,
      93,
      96,
    ],
    "url": "https://jestjs.io",
  },
  Object {
    "count": 2,
    "ids": Array [
      7,
      23,
    ],
    "url": "https://www.respekt.cz",
  },
  Object {
    "count": 2,
    "ids": Array [
      82,
      91,
    ],
    "url": "https://www.valentinog.com",
  },
  Object {
    "count": 2,
    "ids": Array [
      83,
      101,
    ],
    "url": "https://medium.com",
  },
  Object {
    "count": 1,
    "ids": Array [
      5,
    ],
    "url": "https://www.info.cz",
  },
  Object {
    "count": 1,
    "ids": Array [
      65,
    ],
    "url": "https://www.youtube.com",
  },
  Object {
    "count": 1,
    "ids": Array [
      78,
    ],
    "url": "https://github.com",
  },
  Object {
    "count": 1,
    "ids": Array [
      103,
    ],
    "url": "https://www.sitepoint.com",
  },
  Object {
    "count": 1,
    "ids": Array [
      102,
    ],
    "url": "https://shinglyu.com",
  },
]
`;
